### ISSUE TITLE
 in 3.6 import data

### ISSUE THREAD

**Issue by sangouzhi** [JoseConseco/GoB#420]
In Python 3.6, when importing a model from zb to bld, it often becomes invalid and requires clicking the import button again to import the model back into bld.

**Comment by sangouzhi** [comment]
Every time I import a model from ZBrush into BLD3.61, I have to click the import button in BLD for the model to be imported from ZBrush into BLD.

**Comment by kromar** [comment]
Thansk for reporting this issue. Do you have the import button toggled? 

![image](https://github.com/JoseConseco/GoB/assets/1472884/8f76b2cd-cc44-463e-a6c6-5ea43767d6c3)

If so this should update the model in blender whenever you export it form zbrush. If that is not the case its most likely a bug.
Can you make sure you are on the latest version and see if your still have these issues? 
https://github.com/JoseConseco/GoB/releases/tag/V3_5_99

**Comment by sangouzhi** [comment]
thank u for u reply,The problem still exists.
![](https://user-images.githubusercontent.com/1472884/261866005-8f76b2cd-cc44-463e-a6c6-5ea43767d6c3.png)

**Comment by sangouzhi** [comment]
https://github.com/JoseConseco/GoB/assets/114461992/4936704c-14ea-4980-91c2-c938cefea08c

**Comment by sangouzhi** [comment]
https://github.com/JoseConseco/GoB/assets/114461992/67029ea9-b8d3-4986-b25d-6fc475927d98

**Comment by sangouzhi** [comment]
https://github.com/JoseConseco/GoB/assets/114461992/c71dd9da-f892-4b96-87f1-11da106e1d0c

**Comment by sangouzhi** [comment]
> 感谢您报告此问题。您是否已切换导入按钮？
> 
> ![图像](https://user-images.githubusercontent.com/1472884/261866005-8f76b2cd-cc44-463e-a6c6-5ea43767d6c3.png)
> 
> 如果是这样，每当您从 zbrush 导出模型时，都会更新搅拌机中的模型。如果不是这种情况，很可能是一个错误。 您能否确保您使用的是最新版本并查看是否仍然存在这些问题？ https://github.com/JoseConseco/GoB/releases/tag/V3_5_99

Versions 3.4 and up seem to have this problem

**Comment by sangouzhi** [comment]
![Snipaste_2023-08-22_23-30-07](https://github.com/JoseConseco/GoB/assets/114461992/325f9153-6ff1-48ad-a4be-12acc868e672)
The only version that works properly is the one below 3.3, so I've been using that version without upgrading.

**Comment by sangouzhi** [comment]
> 感谢您报告此问题。您是否已切换导入按钮？
> 
> ![图像](https://user-images.githubusercontent.com/1472884/261866005-8f76b2cd-cc44-463e-a6c6-5ea43767d6c3.png)
> 
> 如果是这样，每当您从 zbrush 导出模型时，都会更新搅拌机中的模型。如果不是这种情况，很可能是一个错误。 您能否确保您使用的是最新版本并查看是否仍然存在这些问题？ https://github.com/JoseConseco/GoB/releases/tag/V3_5_99

Before going to bed, I tested it again and for some reason, the issue disappeared this time. The latest version 3.599 is now working perfectly fine on the latest build 3.62, which makes me happy. It seems like the problem has been resolved. I will test it more times and if there are any issues, I will let you know. Thank you, bro!

**Comment by kromar** [comment]
hm i see, sometimes you need to disable and reenable the import toggle to get it working. Looks like there might be some bug that causes it. Btw you can always use the Manual button to import the last object that was exported from zbrush. Maybe you can try that one next time the import fails, it would help narrow it down a bit.

**Comment by sangouzhi** [comment]
> 



> 我明白了，有时您需要禁用并重新启用导入切换才能使其正常工作。看起来可能有一些错误导致了它。顺便说一句，您始终可以使用“手动”按钮导入从 zbrush 导出的最后一个对象。也许您可以在下次导入失败时尝试一下，这将有助于缩小范围。

I tested it again when I got off work today, and it seems like the issue has reappeared. The 'Manual button' can indeed solve the import problem, but it needs to be clicked every time.

**Comment by kromar** [comment]
Do you maybe have any other addons running? I could imagine that something interferes with the import as it runs on a timer and checks for file modifications.

**Comment by sangouzhi** [comment]
> 你可能还有其他插件在运行吗？我可以想象某些东西会干扰导入，因为它在计时器上运行并检查文件修改。

I just deleted all the plugins except for gob. You really guessed it right, buddy. You're amazing, you're my idol. Bro, I need to carefully investigate which plugin caused this error. Thanks again, love you bro, kiss you.

**Comment by kromar** [comment]
Fantastic, let me know if you find out which addon it is for further investigations.

**Comment by sangouzhi** [comment]
> 太棒了，如果您知道它是哪个插件以供进一步调查，请告诉我。

okay, i will

**Comment by kromar** [comment]
> but the smooth groups will be lost

This might be related to the changes that have been done to blender mesh data, i will have a look at this since the smooth groups should be preserved. 
I just checked the past releases and this was introduced back in v3_3_4 so its a bug that should be fixable. https://github.com/JoseConseco/GoB/releases/tag/v3_2_4

Concerning the position and scale, this has been a topic for a long time and i have started some time ago to experiment with some ways to fix it. However ZBrush is a very special piece when it comes to scale and coordinate system and i dont know if its going to be possible to achieve the consistency you are looking for.

To clarify what you want to do, here is a draft of what i understand from your description, it would be great if you could explain it in such a way so i can fully understand the issue you encounter.

1. You send the lowres from blender to zbrush
2. create a high poly model
3. Export the high poly fropm zbrush
4. export the lowres from blender?
5. import both in marmoset?
6. they end up with different positions/scales?

Adding a obj/fbx export would not be that complicated but i would need to understand what the purpose is as this is a bit outside of what GoB should do.

**Comment by sangouzhi** [comment]
> > 但平滑的群体将会丢失
> 
> 这可能与对搅拌机网格数据所做的更改有关，我将看看这一点，因为应该保留平滑组。 我刚刚检查了过去的版本，这是在 v3_3_4 中引入的，所以它是一个应该可以修复的错误。https://github.com/JoseConseco/GoB/releases/tag/v3_2_4
> 
> 关于位置和比例，这已经是一个很长一段时间的话题了，我不久前就开始尝试一些方法来解决它。然而，当涉及到比例和坐标系统时，ZBrush 是一个非常特殊的部分，我不知道它是否能够实现您正在寻找的一致性。
> 
> 为了澄清你想要做什么，这是我从你的描述中理解的草稿，如果你能以这样的方式解释它，这样我就能完全理解你遇到的问题，那就太好了。
> 
> 1. 您将低分辨率从搅拌机发送到 zbrush
> 2. 创建高多边形模型
> 3. 从 zbrush 导出高聚
> 4. 从搅拌机导出低分辨率？
> 5. 将两者都导入狨猴中？
> 6. 他们最终有不同的位置/尺度？
> 
> 添加 obj/fbx 导出不会那么复杂，但我需要了解目的是什么，因为这有点超出了 GoB 应该做的事情。



> > 但平滑的群体将会丢失
> 
> 这可能与对搅拌机网格数据所做的更改有关，我将看看这一点，因为应该保留平滑组。 我刚刚检查了过去的版本，这是在 v3_3_4 中引入的，所以它是一个应该可以修复的错误。https://github.com/JoseConseco/GoB/releases/tag/v3_2_4
> 
> 关于位置和比例，这已经是一个很长一段时间的话题了，我不久前就开始尝试一些方法来解决它。然而，当涉及到比例和坐标系统时，ZBrush 是一个非常特殊的部分，我不知道它是否能够实现您正在寻找的一致性。
> 
> 为了澄清你想要做什么，这是我从你的描述中理解的草稿，如果你能以这样的方式解释它，这样我就能完全理解你遇到的问题，那就太好了。
> 
> 1. 您将低分辨率从搅拌机发送到 zbrush
> 2. 创建高多边形模型
> 3. 从 zbrush 导出高聚
> 4. 从搅拌机导出低分辨率？
> 5. 将两者都导入狨猴中？
> 6. 他们最终有不同的位置/尺度？
> 
> 添加 obj/fbx 导出不会那么复杂，但我需要了解目的是什么，因为这有点超出了 GoB 应该做的事情。

Yes, you understood it correctly. Marmoset Toolbag 4 excels in baking normal and AO maps for models better than any other software. Exporting the model is done specifically for baking these textures. However, I just tested it again in Blender, setting the units correctly and exporting in FBX format, and it seems to work now. The units between Blender and ZBrush appear to be consistent. It's really strange because I tested it multiple times at work in the past few days and it didn't work. It's like every time I talk to you, it brings good luck. Thank you so much! By the way, it seems that the OBJ format always causes problems. I'm not sure if others are aware of this issue. I've attached my Blender export settings. I hope no one else encounters this problem.

![Blender Export Settings](image-link)

**Comment by sangouzhi** [comment]
> > 但平滑的群体将会丢失
> 
> 这可能与对搅拌机网格数据所做的更改有关，我将看看这一点，因为应该保留平滑组。 我刚刚检查了过去的版本，这是在 v3_3_4 中引入的，所以它是一个应该可以修复的错误。https://github.com/JoseConseco/GoB/releases/tag/v3_2_4
> 
> 关于位置和比例，这已经是一个很长一段时间的话题了，我不久前就开始尝试一些方法来解决它。然而，当涉及到比例和坐标系统时，ZBrush 是一个非常特殊的部分，我不知道它是否能够实现您正在寻找的一致性。
> 
> 为了澄清你想要做什么，这是我从你的描述中理解的草稿，如果你能以这样的方式解释它，这样我就能完全理解你遇到的问题，那就太好了。
> 
> 1. 您将低分辨率从搅拌机发送到 zbrush
> 2. 创建高多边形模型
> 3. 从 zbrush 导出高聚
> 4. 从搅拌机导出低分辨率？
> 5. 将两者都导入狨猴中？
> 6. 他们最终有不同的位置/尺度？
> 
> 添加 obj/fbx 导出不会那么复杂，但我需要了解目的是什么，因为这有点超出了 GoB 应该做的事情。

![Snipaste_2023-09-03_23-31-55](https://github.com/JoseConseco/GoB/assets/114461992/af971895-eb2d-4d97-8f68-9df1ff2ce2df)

**Comment by sangouzhi** [comment]
> > 但平滑的群体将会丢失
> 
> 这可能与对搅拌机网格数据所做的更改有关，我将看看这一点，因为应该保留平滑组。 我刚刚检查了过去的版本，这是在 v3_3_4 中引入的，所以它是一个应该可以修复的错误。https://github.com/JoseConseco/GoB/releases/tag/v3_2_4
> 
> 关于位置和比例，这已经是一个很长一段时间的话题了，我不久前就开始尝试一些方法来解决它。然而，当涉及到比例和坐标系统时，ZBrush 是一个非常特殊的部分，我不知道它是否能够实现您正在寻找的一致性。
> 
> 为了澄清你想要做什么，这是我从你的描述中理解的草稿，如果你能以这样的方式解释它，这样我就能完全理解你遇到的问题，那就太好了。
> 
> 1. 您将低分辨率从搅拌机发送到 zbrush
> 2. 创建高多边形模型
> 3. 从 zbrush 导出高聚
> 4. 从搅拌机导出低分辨率？
> 5. 将两者都导入狨猴中？
> 6. 他们最终有不同的位置/尺度？
> 
> 添加 obj/fbx 导出不会那么复杂，但我需要了解目的是什么，因为这有点超出了 GoB 应该做的事情。

Yes, you completely understood. The reason for importing the model into Marmoset Toolbag 4 is because this software has excellent baking capabilities for normal maps, AO maps, and curvature maps, and it's easy to make modifications. Marmoset Toolbag 4 is the best for baking in this regard. Blender's units, orientation, and coordinate system are different from ZBrush's. I tried many times today, and a few times I exported the model from Blender using FBX with units in mm and a scale of 0.001, and the same for ZBrush. But later, when I closed the software and reopened it, it didn't work anymore, and I don't know why.

**Comment by sangouzhi** [comment]
![Snipaste_2023-09-04_00-05-33](https://github.com/JoseConseco/GoB/assets/114461992/5e2a749f-ac18-48f4-a035-01cadca901b7)
Later, I opened Maya and used the GoZ function to export them as FBX files to Marmoset Toolbag. Their size and position were consistent. I hope that Blender and ZBrush can also maintain the same size and position when using GoZ and exporting FBX models.

**Comment by sangouzhi** [comment]
https://github.com/JoseConseco/GoB/assets/114461992/3a1993ed-ea0f-4a4f-9948-dc948218521d

**Comment by sangouzhi** [comment]
https://github.com/JoseConseco/GoB/assets/114461992/5e51eef8-5507-4501-80eb-7f37c2fe1e05

I don't understand how the positions of these two models could have changed. My operations seem to be correct, and I don't understand what's going on.

**Comment by sangouzhi** [comment]
![Snipaste_2023-09-04_00-37-30](https://github.com/JoseConseco/GoB/assets/114461992/dc3f88f4-35a8-479d-9fa7-b57e9c9cb24b)
If both software programs export OBJ files at the same time, rotating the high-poly model 180 degrees on the X-axis will fix the issue. However, this doesn't work with FBX files, and I can't find a solution. But earlier, I tried a few times, and once or twice, I exported the models from both software programs using FBX files, and their positions were completely correct. When I closed the software and opened it again to try again, it was suddenly incorrect again. This is very confusing to me.

**Comment by sangouzhi** [comment]
> 

May I ask if you can take a look at the GoZ plugin code and borrow some of their code? I'm hoping to find some solutions because the units and coordinates of 3ds Max, Maya, and ZBrush are also different, but they export OBJ or FBX files correctly by default. Of course, I don't know if this is feasible, and please forgive me if there's anything wrong with my request.

**Comment by kromar** [comment]
Im sorry i missed these messages somehow. There is a long outstanding issue with transformations which i was trying to solve for some time but thing got in my way. 
This is really tricky with blender and zbrush and its proofing to be a real challenge to fix this issue.

I found out some days ago that i can replace the GoZ export script which i thought was impossible. By replacing this it theoretically allows me to write a custom export script form zbrush which then could use the supported file formats like fbx or obj, however this is a big rewrite of a big part of the addon plus the custom export script which i do not know yet what is possible and there is also not a lot of resources out there for the zscript language which they use so its slow going.
The first thing i want to find out is if the transform issues can be solved with this or if i still need some solution for that when using the fbx format.

**Comment by kromar** [comment]
> > 你可能还有其他插件在运行吗？我可以想象某些东西会干扰导入，因为它在计时器上运行并检查文件修改。
> 
> I just deleted all the plugins except for gob. You really guessed it right, buddy. You're amazing, you're my idol. Bro, I need to carefully investigate which plugin caused this error. Thanks again, love you bro, kiss you.

Did you by any chance find out which addon was causing GoB to fail?

**Comment by sangouzhi** [comment]
> > > 你可能还有其他插件在运行吗？我可以想象某些东西会干扰导入，因为它在计时器上运行并检查文件修改。
> > 
> > 
> > 我刚刚删除了除 gob 之外的所有插件。你真的猜对了，伙计。你太棒了，你是我的偶像。兄弟，我需要仔细调查一下哪个插件导致了这个错误。再次感谢，爱你，兄弟，吻你。
> 
> 您是否有机会发现哪个插件导致 GoB 失败？

The new version 3.60 is completely functional now, thank you.

**Comment by sangouzhi** [comment]
> 很抱歉我错过了这些消息。转换中有一个长期悬而未决的问题，我试图解决这个问题有一段时间了，但事情阻碍了我。 对于搅拌机和 zbrush 来说，这确实很棘手，而且它证明解决这个问题是一个真正的挑战。
> 
> 几天前我发现我可以替换我认为不可能的 GoZ 导出脚本。通过替换它，理论上允许我编写一个来自 zbrush 的自定义导出脚本，然后它可以使用受支持的文件格式，如 fbx 或 obj，但是这是对插件的很大一部分加上自定义导出脚本的重大重写，而我不这样做还不知道什么是可能的，而且他们使用的 zscript 语言也没有太多资源，所以进展缓慢。 我想知道的第一件事是是否可以通过此解决转换问题，或者在使用 fbx 格式时是否仍然需要一些解决方案。

It seems that the export solutions for ZBrush's OBJ and FBX formats are different, and I'm not sure about the specifics. I have tried different settings and units in ZBrush's FBX export, but it seems to be behaving strangely. No matter how I adjust the settings in Blender, I can't match the FBX models exported from ZBrush. So far, the only solution I've found is to use OBJ models, which can be matched by adjusting the settings. However, for my recent projects, which involve small objects, I have been using GoZ to transfer the ZBrush models to Blender for exporting and baking. If you're feeling stuck, it might be a good idea to take a break. Thank you for your efforts and for helping so many people. Sincerely, thank you.

**Comment by kromar** [comment]
I think the main difference in the formats is that FBX supports local transformations while obj only contains global transformations, i would have expected that FBX will be better, especially when the user starts setting custom pivot positions.  That usually fails when using obj and leads to the same transformation issues that the GoZ format has. 
I will do some more consistency testing with FBX as im experimenting with a way to replace the GoZ format.

**Comment by sangouzhi** [comment]
> 我认为格式的主要区别在于 FBX 支持局部变换，而 obj 仅包含全局变换，我预计 FBX 会更好，特别是当用户开始设置自定义枢轴位置时。使用 obj 时通常会失败，并导致与 GoZ 格式相同的转换问题。 我将使用 FBX 进行更多一致性测试，因为我正在尝试替换 GoZ 格式的方法。

"Looking forward to good news."
